fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexScrubbing.java

```diff
@@ -0,0 +1,222 @@
+        return myIndexingTypeStamp;
+    }
+
+    @Nonnull
+    static IndexBuildProto.IndexBuildIndexingStamp compileIndexingTypeStamp() {
+        return
+                IndexBuildProto.IndexBuildIndexingStamp.newBuilder()
+                        .setMethod(IndexBuildProto.IndexBuildIndexingStamp.Method.SCRUB_REPAIR)
+                        .build();
+    }
+
+    @Override
+    CompletableFuture<Void> buildIndexInternalAsync() {
+        return getRunner().runAsync(context ->
+                        context.getReadVersionAsync()
+                                .thenCompose(vignore -> {
+                                    SubspaceProvider subspaceProvider = common.getRecordStoreBuilder().getSubspaceProvider();
+                                    return subspaceProvider.getSubspaceAsync(context)
+                                            .thenCompose(subspace ->
+                                                    indexScrub(subspaceProvider, subspace));
+                                }),
+                common.indexLogMessageKeyValues("IndexScrubbing::buildIndexInternalAsync"));
+    }
+
+    @Nonnull
+    private CompletableFuture<Void> indexScrub(@Nonnull SubspaceProvider subspaceProvider, @Nonnull Subspace subspace) {
+
+        final List<Object> additionalLogMessageKeyValues = Arrays.asList(LogMessageKeys.CALLING_METHOD, "indexScrub");
+
+        return iterateAllRanges(additionalLogMessageKeyValues,
+                this::indexScrubRangeOnly,
+                subspaceProvider, subspace);
+    }
+
+    @Nonnull
+    private CompletableFuture<Boolean> indexScrubRangeOnly(@Nonnull FDBRecordStore store, @Nonnull AtomicLong recordsScanned) {
+        // return false when done
+        Index index = common.getIndex();
+        final RecordMetaData metaData = store.getRecordMetaData();
+        final RecordMetaDataProvider recordMetaDataProvider = common.getRecordStoreBuilder().getMetaDataProvider();
+        if (recordMetaDataProvider == null || !metaData.equals(recordMetaDataProvider.getRecordMetaData())) {
+            throw new MetaDataException("Store does not have the same metadata");
+        }
+        final IndexMaintainer maintainer = store.getIndexMaintainer(index);
+        final IndexScrubbingTools<?> tools = maintainer.getIndexScrubbingTools(scrubbingType);
+        if (tools == null) {
+            throw new UnsupportedOperationException("This index does not support scrubbing type " + scrubbingType);
+        }
+
+        return indexScrubRangeOnly(store, recordsScanned, index, tools, maintainer.isIdempotent());
+    }
+
+    private <T> CompletableFuture<Boolean> indexScrubRangeOnly(final @Nonnull FDBRecordStore store, final @Nonnull AtomicLong recordsScanned, final Index index, final IndexScrubbingTools<T> tools, boolean isIdempotent) {
+        // scrubbing only readable
+        validateOrThrowEx(store.getIndexState(index) == IndexState.READABLE, "scrubbed index is not readable");
```

Comment:
👍 
It can surely be `READABLE_UNIQUE_PENDING` (not the same rangeSet). Supporting that should probably be done in another PR as it requires some changes in the `IndexingBase` + testing.